queue/retry.go

```diff
@@ -11,12 +11,36 @@ import (
 	"github.com/mongodb/grip/message"
 	"github.com/mongodb/grip/recovery"
 	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/mongo"
 )
 
+// statsRetryHandler is the same as an amboy.RetryHandler but allows it to
+// provide additional runtime statistics on its jobs. This is primarily for
+// queue-internal tracking and testing purposes.
+type statsRetryHandler interface {
+	amboy.RetryHandler
+	// stats returns information about the job statistics for an
+	// amboy.RetryHandler. Results are reset with each runtime execution.
+	stats() retryHandlerStats
+}
+
+// retryHandlerStats captures runtime statistics on an amboy.RetryHandler.
+type retryHandlerStats struct {
+	pending   int
+	completed int
+	retried   int
+}
+
+type retryingJobOutcome struct {
+	job     amboy.RetryableJob
+	retried bool
+}
+
 type basicRetryHandler struct {
 	queue         amboy.RetryableQueue
 	opts          amboy.RetryHandlerOptions
 	pending       map[string]amboy.RetryableJob
+	completed     map[string]retryingJobOutcome
```

Comment:
Will holding pending and completed jobs in memory cause issues?

Reply:
Same comment about temporary stats.